### PR TITLE
fix: Missing dependencies / setup fails on new installations

### PR DESCRIPTION
**Pull request by pmzandbergen**
Missing dependency may cause new installations to fail.

See: https://github.com/Adyen/adyen-magento2/issues/1431#issuecomment-1941546379

Fixes:
- #2526 
- #1431

### CODE REVIEW

**Review by candemiralp — APPROVED** [review]

**Review by hossam-adyen — APPROVED** [review]

**Comment by pmzandbergen** [comment]
@hossam-adyen these problems were introduced by: https://github.com/Adyen/adyen-magento2/pull/2476

@bramstroker will add the details

**Comment by bramstroker** [comment]
Narrowed down the issue to be in https://github.com/Adyen/adyen-magento2/blob/develop/Setup/Recurring.php

During execution of the recurring schema the following components will be constructed by DI container:

- \Adyen\Payment\Setup\Recurring
- \Adyen\Payment\Helper\PaymentMethodsFactory
- \Adyen\Payment\Helper\Data
- \Magento\Tax\Model\Calculation
- \Magento\Customer\Model\Session
- \Magento\Customer\Model\ResourceModel\Customer

Construction of Customer resource calls `setType` and fails with following exception, thus crashing Magento installation.

`Invalid entity_type specified: customer`

The eav entity types are created by data patch in `Magento_Customer` module, and data patches are being executed after recurring schema.

The question is why this is in recurring schema in the first place as it has nothing to do with DB schema. It just sets some config settings.

So the solution seems to me to execute this logic in a data patch, which ultimately also has a dependency on the Magento data patch which creates the `eav_entity_type` records. `\Magento\Customer\Setup\Patch\Data\DefaultCustomerGroupsAndAttributes`

**Comment by hossam-adyen** [comment]
Hi @pmzandbergen, thanks for your valuable contribution and dedicate time for opening this draft and providing this fix, please let us know when it's ready for review.

If you also don't mine, please update [the unit test file](https://github.com/Adyen/adyen-magento2/blob/develop/Test/Unit/Console/Command/WebhookProcessorCommandTest.php) to avoid the failure of the PR checks

**Comment by pmzandbergen** [comment]
> Hi @pmzandbergen, thanks for your valuable contribution and dedicate time for opening this draft and providing this fix, please let us know when it's ready for review.
> 
> If you also don't mine, please update [the unit test file](https://github.com/Adyen/adyen-magento2/blob/develop/Test/Unit/Console/Command/WebhookProcessorCommandTest.php) to avoid the failure of the PR checks

Have a look at: https://github.com/Adyen/adyen-magento2/pull/2509#issuecomment-1941717479

**Comment by bramstroker** [comment]
I have changed current `Recurring` schema script, to `RecurringData`, which solves our problems.

Workflow for CI/CD testing pipeline in our projects is as follows:
- `composer install` using existing composer.lock file, containing both magento installation and 3rd party modules (also the adyen module).
- `bin/magento setup:install`

This fails without the fixes provided in this PR.

It would be nice if similar procedure also can be verified in you QA pipeline, as it's not the first time our CI/CD pipelines break after (minor) upgrade of Adyen module.

**Review by candemiralp — DISMISSED** [review]

File: Setup/RecurringData.php [hunk 1]
```diff
@@ -4,11 +4,12 @@
...
 use Adyen\Payment\Helper\PaymentMethods;
 use Adyen\Payment\Helper\PaymentMethodsFactory;
-use Magento\Framework\Setup\InstallSchemaInterface;
+use Magento\Framework\Setup\InstallDataInterface;
 use Magento\Framework\Setup\ModuleContextInterface;
+use Magento\Framework\Setup\ModuleDataSetupInterface;
 use Magento\Framework\Setup\SchemaSetupInterface;
 
-class Recurring implements InstallSchemaInterface
+class RecurringData implements InstallDataInterface
```
Comment (hossam-adyen) [inline Setup/RecurringData.php:12]: Please update the related unit test file as it give the below error due to the name change

```
1) Adyen\Payment\Test\Unit\Setup\RecurringTest::testInstall
Error: Class "Adyen\Payment\Setup\Recurring" not found

/var/www/html/vendor/adyen/module-payment/Test/Unit/Setup/RecurringTest.php:29

```

Comment (bramstroker) [reply Setup/RecurringData.php:12]: Should be fixed now.
Screenshot 2024-02-19 at 16 23 50

**Review by hossam-adyen — APPROVED** [review]

**Review by candemiralp — APPROVED** [review]

**Comment by bramstroker** [comment]
Can this be merged please so it can be part of upcoming release?
This currently blocks updates for the module for our clients.
Don't like to put any temporary composer patches to work around the issue.

**Pull request merged**